src/mame/konami/konamigv.cpp

```diff
@@ -680,19 +779,115 @@ void konamigv_state::tokimeki_serial_w(offs_t offset, uint16_t data, uint16_t me
+	{
+		m_serial_sensor_data = 0;
+		m_serial_sensor_id = 0;
+		m_serial_val = 0;
+		m_serial_len = 0;
+	}
+	else if (!m_serial_clk && clk)
+	{
+		if (m_serial_len < 5)
+		{
+			// Sends 5 bits of data for the sensor ID
+			m_serial_val |= BIT(data, 4) << (4 - m_serial_len);
+
+			if (m_serial_len == 4)
+			{
+				m_serial_sensor_id = m_serial_val;
+
+				switch (m_serial_sensor_id)
+				{
+					case 0x1a: // GSR sensor value
+						m_serial_sensor_data = m_gsr->read();
+						break;
+					case 0x18:
+					case 0x19:
+					default:
+						m_serial_sensor_data = 0;
+						break;
+				}
+			}
+		}
+		else if (m_serial_len >= 6)
+		{
+			// Shifts data between reads of tokimeki_serial_r
+			m_serial_sensor_data <<= 1;
+		}
+
+		m_serial_len++;
+	}
 
+	m_serial_clk = clk;
 }
 
-void konamigv_state::tmosh(machine_config &config)
+void tokimeki_state::tmosh(machine_config &config)
 {
 	konamigv(config);
-	m_maincpu->set_addrmap(AS_PROGRAM, &konamigv_state::tmosh_map);
+	m_maincpu->set_addrmap(AS_PROGRAM, &tokimeki_state::tmosh_map);
 }
 
+void tokimeki_state::tmoshs_init()
+{
+	m_device_val_start_state = 0xf073;
+}
+
+void tokimeki_state::tmoshsp_init()
+{
+	m_device_val_start_state = 0xf0ba;
+}
+
+CUSTOM_INPUT_MEMBER(tokimeki_state::tokimeki_device_check_r)
+{
+	return BIT(m_device_val, 15);
+}
+
+void tokimeki_state::tokimeki_device_check_w(int state)
+{
+	// The check is accepted when it reads whatever is specified in m_device_val_start_state
+	if (state)
+		m_device_val = (m_device_val << 1) | BIT(m_device_val, 15);
+}
+
+static INPUT_PORTS_START( tmosh )
+	PORT_INCLUDE( konamigv )
+
+	PORT_MODIFY("P2")
+	PORT_BIT( 0x00000400, IP_ACTIVE_HIGH, IPT_CUSTOM ) PORT_CUSTOM_MEMBER(tokimeki_state, tokimeki_device_check_r)
+
+	PORT_MODIFY("EEPROMOUT")
+	PORT_BIT( 0x10, IP_ACTIVE_HIGH, IPT_OUTPUT ) PORT_WRITE_LINE_MEMBER(tokimeki_state, tokimeki_device_check_w)
+
+	// valid range for heart rate is 50-150, anything outside of that range makes the game show ? in the heart rate meter area
+	// Setting the value here to 0 will act as if the player's hand is off the sensor, and anything after that acts as 50-100
+	// Default heartbeat is configured for 80 beats per minute
+	PORT_START("HEARTBEAT")
+	PORT_BIT( 0xff, 31, IPT_PADDLE_V ) PORT_SENSITIVITY(10) PORT_KEYDELTA(10) PORT_CENTERDELTA(0) PORT_MINMAX(0, 101) PORT_NAME("Heartbeat Sensor")
+
+	// value read during calibration is treated as zero
+	// scale in operator menu goes from 0x00-0xff but only 0x00-0x80 is actually usable in-game
+	PORT_START("GSR")
+	PORT_BIT( 0xff, 0x00, IPT_POSITIONAL ) PORT_SENSITIVITY(100) PORT_KEYDELTA(2) PORT_CENTERDELTA(0) PORT_MINMAX(0x00, 0x80) PORT_NAME("GSR Sensor")
+
+	PORT_START("PRINTER_META")
+	PORT_DIPNAME( 0x01, 0x01, "Printer Paper Empty" )
+	PORT_DIPSETTING( 0x01, DEF_STR( Off ) )
+	PORT_DIPSETTING( 0x00, DEF_STR( On ) )
+	PORT_DIPNAME( 0x02, 0x02, "Printer Malfunction" )
+	PORT_DIPSETTING( 0x02, DEF_STR( Off ) )
+	PORT_DIPSETTING( 0x00, DEF_STR( On ) )
```

Comment:
Please use either `IPT_OTHER` with `PORT_TOGGLE` or `PORT_CONFNAME` etc. for these as they aren’t real DIP switches.

Reply:
Changed it to `PORT_CONFNAME` because I don't really want them showing up as keybinds like what happens with `IPT_OTHER`.